### ISSUE TITLE
[addon-docs] DocGen description are not displayed with StyledComponent

### ISSUE THREAD

**Issue by davidouanounou** [storybookjs/storybook#8279]
**Describe the bug**
Description related to the component or to the props are not displayed when component is based on StyledComponent.

**To Reproduce**
1. Create a component based on StyledComponent
2. Launch Storybook

**Expected behavior**
Should display description as expected

**Screenshots**
![image](https://user-images.githubusercontent.com/1861711/66110856-eed99400-e5c7-11e9-9f13-f384c08edad8.png)

**Code snippets**
import styled from 'styled-components'
import PropTypes from 'prop-types';

/**
 * Use `A` to provide a regular link
 */
const A = styled('a')(
    {
        margin: '8px 0',
        outline: 'none'
    }
)

A.displayName = 'Link';
A.defaultProps = {
    children: 'This is a link'
};

A.propTypes = {
    /** That should be the clickable element */
    children: PropTypes.node.isRequired
};
export default A;

**System:**
System:
    OS: macOS Mojave 10.14.6
    CPU: (8) x64 Intel(R) Core(TM) i7-8569U CPU @ 2.80GHz
  Binaries:
    Node: 12.4.0 - /usr/local/bin/node
    Yarn: 1.17.0 - /usr/local/bin/yarn
    npm: 6.11.3 - /usr/local/bin/npm
  Browsers:
    Chrome: 77.0.3865.90
    Safari: 13.0.1
  npmPackages:
    @storybook/addon-actions: ^5.2.1 => 5.2.1 
    @storybook/addon-docs: ^5.3.0-alpha.6 => 5.3.0-alpha.6 
    @storybook/addon-info: ^5.2.1 => 5.2.1 
    @storybook/addon-knobs: ^5.2.1 => 5.2.1 
    @storybook/addon-links: ^5.2.1 => 5.2.1 
    @storybook/addon-storysource: ^5.2.1 => 5.2.1 
    @storybook/addons: ^5.2.1 => 5.2.1 
    @storybook/react: ^5.2.1 => 5.2.1 

**Additional context**
Nothing special

**Comment by salmanm** [comment]
I think this is related to https://github.com/storybookjs/babel-plugin-react-docgen/pull/74.

Can we get some feedback on that PR?

**Comment by KoshiroKun** [comment]
In 5.2.4 I see the props from styled components in some components, not in all. Idk if it's my fault or not but I have all components styled and with stories in the same way.

Can you confirm this issue is resolved on 5.2.4? Thanks.

**Comment by michaelcitro** [comment]
Hi @shilman I am having this issue as well but in my case, it seems to happen when I am deconstructing the rest of the props in my component. For example:

```
const Input = (props) => {
  const { prop1, prop2, ...rest } = props;
  return <input {...rest} />;
};

Input.propTypes = {
  prop1: PropTypes.string,
  prop2: PropTypes.string,
  prop3: PropTypes.string,
};

export default Input;
```

when I remove `...rest` or deconstruct each prop indiviguallty it works fine though.

**Comment by shilman** [comment]
@michaelcitro thanks for the info. i'm making a bunch of repros now for all of the issues raised in #8435 (umbrella issue, that includes this one) and then will go through & diagnose. appreciate your patience!

**Comment by shilman** [comment]
Repro: https://github.com/storybookjs/storybook/commit/eaa147ae6239533c6db352e97cf06946c2720c65

**Comment by maeriens** [comment]
Not sure if this adds much - but I am able to show the docGen if I create a React functional component that returns the Styled Component. But it means wrapping all the styled components that way...
```jsx
mport React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const Anchor = styled.a`
  color: teal;
  text-decoration: none;
`;

/** Now this would `work` but it's not the best approach */
const NewAnchor = (props) => <Anchor {...props}/>

NewAnchor.propTypes = {
  /** Show me your color */
  color: PropTypes.string
}

export default NewAnchor;
```

Proof: 
![image](https://user-images.githubusercontent.com/5074186/80845491-be022180-8bdf-11ea-916d-5611377da154.png)

All storybook versions are `5.3.18`. Styled Components `4.4..1`.

**Comment by testacode** [comment]
> Not sure if this adds much - but I am able to show the docGen if I create a React functional component that returns the Styled Component. But it means wrapping all the styled components that way...
> 
> ```js-jsx
> import React from 'react';
> import styled from 'styled-components';
> import PropTypes from 'prop-types';
> 
> const Anchor = styled.a`
>   color: ${props => props.color};
>   text-decoration: none;
> `;
> 
> /** Now this would `work` but it's not the best approach */
> const NewAnchor = (props) => <Anchor {...props}/>
> 
> NewAnchor.propTypes = {
>   /** Show me your color */
>   color: PropTypes.string
> }
> 
> NewAnchor.defaultProps = { color: teal };
> 
> export default NewAnchor;
> ```
> 
> Proof:
> ![image](https://user-images.githubusercontent.com/5074186/80845491-be022180-8bdf-11ea-916d-5611377da154.png)
> 
> All storybook versions are `5.3.18`. Styled Components `4.4.1`.

I personally don't like to change my Components so that they can work with Storybook. It's like changing code for making tests pass...

**Comment by shilman** [comment]
@testacode that's why it's a workaround, not a proper fix

**Comment by maeriens** [comment]
@shilman to confirm - the issue here is in `react-docgen`, right? Haven't investigated much since but it seems not related to Storybook

**Comment by shilman** [comment]
Yeah, this particular codepath (react PropTypes) would get handled by `react-docgen` @maeriens

**Comment by penx** [comment]
I'm pretty sure this is because of https://github.com/reactjs/react-docgen/issues/256

There is a custom resolver https://github.com/jquense/react-docgen-styled-resolver

But this only picks up use of `styled` with a template literal and not when used as `styled({ someStyles })`

https://github.com/jquense/react-docgen-styled-resolver/issues/1

We encountered issues with the workaround mentioned above in https://github.com/govuk-react/govuk-react/issues/527. Unfortunately it was a few years ago so can't remember the specifics. We resolved this by exporting the `const DocumentedComponent = (props) => <Component {...props}/>` separately to the main component and using this export when generating docs.

I would love a real fix for this :D

**Comment by penx** [comment]
@shilman any thoughts on this? Perhaps allowing Storybook's docgen to be configured with a custom resolver would help for some people?

**Comment by maeriens** [comment]
Chiming in - from the issue linked by the Storybook docs itself (see [here](https://github.com/storybookjs/storybook/tree/next/addons/docs/react#typescript-props-with-react-docgen)), in this [comment](https://github.com/reactjs/react-docgen/issues/387#issuecomment-602803231), @Winner95 explains he made another resolver [Typescript-react-function-component-props-handler](https://github.com/Winner95/typescript-react-function-component-props-handler) that even has an example running with the beta of [Storybook 6](https://codesandbox.io/s/initial-600-beta22-config-working-prototype-with-showing-props-react-docgen-ftvf2).

**Comment by penx** [comment]
Unfortunately for our use case, we're not "generating props" but need something that pulls out the docgen comment from TypeScript for inclusion in storybook.

I switched our project over to using TypeScript this morning (still using prop types for now) and noticed none of the docgen comments were appearing so had to switch to using this config in .storybook/main.js

```js
  typescript: {
    reactDocgen: 'react-docgen',
  }
```

It's good to see that it's possible to use a custom docgen resolver with Storybook, but from what I can see, the solution is to put this in .storybook/main.js:


```js
  webpackFinal: async config => {
    const reactDocgen = [
      "babel-plugin-react-docgen",
      {
        DOC_GEN_COLLECTION_NAME: "STORYBOOK_REACT_CLASSES",
        // custom handlers should go before the default ones
        preHandlers: ["typescript-react-function-component-props-handler"]
      }
    ];

    config.module.rules[3].oneOf[1].options.plugins.push(reactDocgen);

    return config;
  }
```

I expect this would be fairly fragile as-is so would be good to have a direct option for a custom resolver from the Storybook config, if that's seen as a possible way forwards.

**Comment by shilman** [comment]
@penx we currently recommend `react-docgen-typescript` for processing typescript files. When we fully support `react-docgen`, we could look into stylized ways of supporting better configuration.

**Comment by siminino** [comment]
@shilman Now do we have stylized way to configure react-docgen with a custom resolver?
If not, do we have any documentation about how is the best way to do it?

**Comment by valentinpalkovic** [comment]
This was reported on Storybook 5.3-alpha against a docgen pipeline that has been substantially rewritten (React default extractor switched from `react-docgen-typescript` to `react-docgen` in v8; ArgsTable → ArgTypes block; new addon-docs blocks split). The `has workaround` label is set and there's been no engineering signal in 3+ years. Closing — please open a fresh issue with a repro on 10.x if it still reproduces.
